server/src-lib/Hasura/RQL/DDL/Remote/Types.hs

```diff
@@ -0,0 +1,157 @@
+    , rmfParamMap :: !(HashMap G.Name InpValInfo)
+    }
+  deriving (Show, Eq, Lift)
+
+data RemoteRelationship =
+  RemoteRelationship
+    { rtrName :: RemoteRelationshipName
+    , rtrTable :: QualifiedTable
+    , rtrRemoteSchema :: RemoteSchemaName
+    , rtrRemoteField :: G.Name
+    , rtrHasuraFields :: Set FieldName
+    , rtrRemoteArguments :: RemoteArguments
+    }
+  deriving (Show, Eq, Lift)
+
+newtype RemoteSchemaName
+  = RemoteSchemaName
+  { unRemoteSchemaName :: Text}
+  deriving (Show, Eq, Lift, Hashable, ToJSON, ToJSONKey, FromJSON, Q.ToPrepArg, Q.FromCol)
+
+newtype RemoteRelationshipName
+  = RemoteRelationshipName
+  { unRemoteRelationshipName :: Text}
+  deriving (Show, Eq, Lift, Hashable, ToJSON, ToJSONKey, FromJSON, Q.ToPrepArg, Q.FromCol)
+
+$(deriveJSON (aesonDrop 3 snakeCase) ''RemoteRelationship)
+
+--------------------------------------------------------------------------------
+-- Custom JSON roundtrip instances for RemoteField and down
+
+instance ToJSON RemoteField where
+  toJSON RemoteField {..} =
+    object
+      [ "remote_relationship" .= toJSON rmfRemoteRelationship
+      , "g_type" .= toJsonGType rmfGType
+      , "param_map" .= fmap toJsonInpValInfo rmfParamMap
+      ]
+
+instance FromJSON RemoteField where
+  parseJSON value = do
+    hmap <- parseJSON value
+    rmfRemoteRelationship <- hmap .: "remote_relationship"
+    rmfGType <- hmap .: "g_type" >>= parseJsonGType
+    rmfParamMap <- hmap .: "param_map" >>= traverse parseJsonInpValInfo
+    pure RemoteField {..}
+
+-- | Parse a GType, using Either as an auxilliary type.
+parseJsonGType :: Value -> Parser G.GType
+parseJsonGType value = do
+  oneof <- parseJSON value
+  case oneof of
+    Left (nullability, namedType) ->
+      pure (G.TypeNamed (G.Nullability nullability) namedType)
+    Right (nullability, listTypeValue) -> do
+      listType <- fmap G.ListType (parseJsonGType listTypeValue)
+      pure (G.TypeList (G.Nullability nullability) listType)
+
+-- | Convert to JSON, using Either as an auxilliary type.
+toJsonGType :: G.GType -> Value
+toJsonGType gtype =
+  toJSON
+    (case gtype of
+       G.TypeNamed (G.Nullability nullability) namedType ->
+         Left (nullability, namedType)
+       G.TypeList (G.Nullability nullability) (G.ListType listType) ->
+         Right (nullability, listType))
+
+parseJsonInpValInfo :: Value -> Parser InpValInfo
+parseJsonInpValInfo value = do
+  hashmap <- parseJSON value
+  _iviDesc <- hashmap .: "desc"
+  _iviName <- hashmap .: "name"
+  _iviDefVal <-
+    hashmap .: "def_val" >>= maybe (pure Nothing) (fmap Just . parseValueConst)
+  _iviType <- hashmap .: "type" >>= parseJsonGType
+  pure InpValInfo {..}
+
+toJsonInpValInfo :: InpValInfo -> Value
+toJsonInpValInfo InpValInfo {..} =
+  object
+    [ "desc" .= _iviDesc
+    , "name" .= _iviName
+    , "def_val" .= fmap gValueConstToValue _iviDefVal
+    , "type" .= _iviType
+    ]
+
+gValueConstToValue :: G.ValueConst -> A.Value
```

Comment:
can we not use https://github.com/hasura/graphql-engine/blob/a8dd3b3deb11b41300fa7a821ac05037c424282b/server/src-lib/Hasura/GraphQL/Validate/InputValue.hs#L147 here?

Reply:
You could, they are the same.